File: starrysky/src/main/java/com/lzx/starrysky/provider/MediaQueueProviderImpl.java
package com.lzx.starrysky.provider;

import android.graphics.Bitmap;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.text.TextUtils;

import com.lzx.starrysky.BaseMediaInfo;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;


/**
 * 媒体信息提供类
 */
public class MediaQueueProviderImpl implements MediaQueueProvider {

    //使用Map在查找方面会效率高一点
    private LinkedHashMap<String, BaseMediaInfo> mediaListMap;
    private LinkedHashMap<String, MediaMetadataCompat> mMediaMetadataCompatMap;
    private List<BaseMediaInfo> mediaList;
    private LinkedHashMap<String, SongInfo> songListMap;
    private List<SongInfo> songList;

    public MediaQueueProviderImpl() {
        mediaListMap = new LinkedHashMap<>();
        mMediaMetadataCompatMap = new LinkedHashMap<>();
        songListMap = new LinkedHashMap<>();
        mediaList = new ArrayList<>();
        songList = new ArrayList<>();
    }

    @Override
    public List<BaseMediaInfo> getMediaList() {
        return mediaList;
    }

    @Override
    public List<SongInfo> getSongList() {
        return songList;
    }

    @Override
    public void updateMediaList(List<BaseMediaInfo> mediaInfoList) {
        mediaList.clear();
        mediaListMap.clear();
        mediaList.addAll(mediaInfoList);
        for (BaseMediaInfo info : mediaInfoList) {
            mediaListMap.put(info.getMediaId(), info);
        }
    }

    @Override
    public void updateMediaListBySongInfo(List<SongInfo> songInfos) {
        songListMap.clear();
        mMediaMetadataCompatMap.clear();
        songList.clear();
        songList.addAll(songInfos);
        List<BaseMediaInfo> mediaInfos = new ArrayList<>();
        for (SongInfo songInfo : songInfos) {
            BaseMediaInfo mediaInfo = new BaseMediaInfo();
            mediaInfo.setMediaId(songInfo.getSongId());
            mediaInfo.setMediaTitle(songInfo.getSongName());
            mediaInfo.setMediaCover(songInfo.getSongCover());
            mediaInfo.setMediaUrl(songInfo.getSongUrl());
            mediaInfo.setDuration(songInfo.getDuration());
            mediaInfos.add(mediaInfo);
            songListMap.put(songInfo.getSongId(), songInfo);
        }
        mMediaMetadataCompatMap = toMediaMetadata(songInfos);
        updateMediaList(mediaInfos);
    }

    @Override
    public void addMediaInfo(BaseMediaInfo mediaInfo) {
        if (mediaInfo == null) {
            return;
        }
        if (!mediaList.contains(mediaInfo)) {
            mediaList.add(mediaInfo);
        }
        mediaListMap.put(mediaInfo.getMediaId(), mediaInfo);
    }

    @Override
    public boolean hasMediaInfo(String songId) {
        return mediaListMap.containsKey(songId);
    }

    @Override
    public BaseMediaInfo getMediaInfo(String songId) {
        if (TextUtils.isEmpty(songId)) {
            return null;
        }
        if (hasMediaInfo(songId)) {
            return mediaListMap.get(songId);
        } else {
            return null;
        }
    }

    @Override
    public BaseMediaInfo getMediaInfo(int index) {
        if (index < 0 || index >= mediaList.size()) {
            return null;
        }
        return mediaList.get(index);
    }

    @Override
    public SongInfo getSongInfo(int index) {
        if (index < 0 || index >= songList.size()) {
            return null;
        }
        return songList.get(index);
    }

    @Override
    public SongInfo getSongInfo(String songId) {
        if (TextUtils.isEmpty(songId)) {
            return null;
        }
        if (songListMap.containsKey(songId)) {
            return songListMap.get(songId);
        } else {
            return null;
        }
    }

    @Override
    public int getIndexByMediaId(String songId) {
        BaseMediaInfo info = getMediaInfo(songId);
        return info != null ? getMediaList().indexOf(info) : -1;
    }

    /**
     * 获取List#MediaMetadataCompat
     */
    @Override
    public List<MediaMetadataCompat> getMediaMetadataCompatList() {
        return new ArrayList<>(mMediaMetadataCompatMap.values());
    }

    /**
     * 获取 List#MediaBrowserCompat.MediaItem 用于 onLoadChildren 回调
     */
    @Override
    public List<MediaBrowserCompat.MediaItem> getChildrenResult() {
        List<MediaBrowserCompat.MediaItem> mediaItems = new ArrayList<>();
        List<MediaMetadataCompat> list = getMediaMetadataCompatList();
        for (MediaMetadataCompat metadata : list) {
            MediaBrowserCompat.MediaItem mediaItem = new MediaBrowserCompat.MediaItem(
                    metadata.getDescription(),
                    MediaBrowserCompat.MediaItem.FLAG_PLAYABLE);
            mediaItems.add(mediaItem);
        }
        return mediaItems;
    }

    /**
     * 获取乱序列表
     */
    @Override
    public Iterable<MediaMetadataCompat> getShuffledMediaMetadataCompat() {
        List<MediaMetadataCompat> shuffled = new ArrayList<>(mMediaMetadataCompatMap.values());
        Collections.shuffle(shuffled);
        return shuffled;
    }

    @Override
    public Iterable<BaseMediaInfo> getShuffledMediaInfo() {
        Collections.shuffle(mediaList);
        return mediaList;
    }

    /**
     * 根据id获取对应的MediaMetadataCompat对象
     */
    @Override
    public MediaMetadataCompat getMediaMetadataCompatById(String songId) {
        return mMediaMetadataCompatMap.containsKey(songId) ? mMediaMetadataCompatMap.get(songId) : null;
    }

    /**
     * 更新封面art
     */
    @Override
    public synchronized void updateMusicArt(String songId, MediaMetadataCompat changeData, Bitmap albumArt, Bitmap icon) {
        MediaMetadataCompat metadata = new MediaMetadataCompat.Builder(changeData)
                .putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, albumArt)
                .putBitmap(MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON, icon)
                .build();
        mMediaMetadataCompatMap.put(songId, metadata);
    }

    /**
     * List<SongInfo> 转 ConcurrentMap<String, MediaMetadataCompat>
     */
    private synchronized static LinkedHashMap<String, MediaMetadataCompat> toMediaMetadata(List<SongInfo> songInfos) {
        LinkedHashMap<String, MediaMetadataCompat> map = new LinkedHashMap<>();
        for (SongInfo info : songInfos) {
            MediaMetadataCompat metadataCompat = toMediaMetadata(info);
            map.put(info.getSongId(), metadataCompat);
        }
        return map;
    }

    /**
     * SongInfo 转 MediaMetadataCompat
     */
    private synchronized static MediaMetadataCompat toMediaMetadata(SongInfo info) {
        String albumTitle = "";
        if (!TextUtils.isEmpty(info.getAlbumName())) {
            albumTitle = info.getAlbumName();
        } else if (!TextUtils.isEmpty(info.getSongName())) {
            albumTitle = info.getSongName();
        }
        String songCover = "";
        if (!TextUtils.isEmpty(info.getSongCover())) {
            songCover = info.getSongCover();
        } else if (!TextUtils.isEmpty(info.getAlbumCover())) {
            songCover = info.getAlbumCover();
        }
        MediaMetadataCompat.Builder builder = new MediaMetadataCompat.Builder();
        builder.putString(MediaMetadataCompat.METADATA_KEY_MEDIA_ID, info.getSongId());
        builder.putString(MediaMetadataCompat.METADATA_KEY_MEDIA_URI, info.getSongUrl());
        if (!TextUtils.isEmpty(albumTitle)) {
            builder.putString(MediaMetadataCompat.METADATA_KEY_ALBUM, albumTitle);
        }
        if (!TextUtils.isEmpty(info.getArtist())) {
            builder.putString(MediaMetadataCompat.METADATA_KEY_ARTIST, info.getArtist());
        }
        if (info.getDuration() != -1) {
            builder.putLong(MediaMetadataCompat.METADATA_KEY_DURATION, info.getDuration());
        }
        if (!TextUtils.isEmpty(info.getGenre())) {
            builder.putString(MediaMetadataCompat.METADATA_KEY_GENRE, info.getGenre());
        }
        if (!TextUtils.isEmpty(songCover)) {
            builder.putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI, songCover);
        }
        if (!TextUtils.isEmpty(info.getSongName())) {
            builder.putString(MediaMetadataCompat.METADATA_KEY_TITLE, info.getSongName());
        }
        if (info.getTrackNumber() != -1) {
            builder.putLong(MediaMetadataCompat.METADATA_KEY_TRACK_NUMBER, info.getTrackNumber());
        }
        builder.putLong(MediaMetadataCompat.METADATA_KEY_NUM_TRACKS, info.getAlbumSongCount());
        return builder.build();
    }
}

File: starrysky/src/main/java/com/lzx/starrysky/provider/MediaQueueProviderImpl.kt
package com.lzx.starrysky.provider

import android.graphics.Bitmap
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import android.text.TextUtils
import com.lzx.starrysky.BaseMediaInfo
import com.lzx.starrysky.ext.album
import com.lzx.starrysky.ext.albumArtUri
import com.lzx.starrysky.ext.artist
import com.lzx.starrysky.ext.duration
import com.lzx.starrysky.ext.genre
import com.lzx.starrysky.ext.id
import com.lzx.starrysky.ext.mediaUri
import com.lzx.starrysky.ext.title
import com.lzx.starrysky.ext.trackCount
import com.lzx.starrysky.ext.trackNumber
import com.lzx.starrysky.utils.StarrySkyUtils

open class MediaQueueProviderImpl : MediaQueueProvider {

    private var mediaListMap = linkedMapOf<String, BaseMediaInfo>()
    private var mMediaMetadataCompatMap = linkedMapOf<String, MediaMetadataCompat>()
    private var songListMap = linkedMapOf<String, SongInfo>()

    private val mediaList = mutableListOf<BaseMediaInfo>()
    private val songList = mutableListOf<SongInfo>()

    override val mediaMetadataCompatList: List<MediaMetadataCompat>
        get() = mediaMetadataCompatList.toList()

    override val childrenResult: List<MediaBrowserCompat.MediaItem>
        get() {
            val mediaItems = mutableListOf<MediaBrowserCompat.MediaItem>()
            for (metadata in mediaMetadataCompatList) {
                val mediaItem = MediaBrowserCompat.MediaItem(
                    metadata.description,
                    MediaBrowserCompat.MediaItem.FLAG_PLAYABLE)
                mediaItems.add(mediaItem)
            }
            return mediaItems
        }


    override val shuffledMediaMetadataCompat: Iterable<MediaMetadataCompat>
        get() {
            val shuffled = ArrayList(mMediaMetadataCompatMap.values)
            shuffled.shuffle()
            return shuffled
        }

    override val shuffledMediaInfo: Iterable<BaseMediaInfo>
        get() {
            mediaList.shuffle()
            return mediaList
        }

    override fun getMediaList(): MutableList<BaseMediaInfo> {
        StarrySkyUtils.log("mediaList = " + mediaList.size)
        return mediaList
    }

    override fun getSongList(): MutableList<SongInfo> {
        return songList
    }

    override fun updateMediaList(mediaInfoList: List<BaseMediaInfo>) {
        //更新 mediaList，mediaListMap
        mediaList.clear()
        mediaList.addAll(mediaInfoList)
        for (info in mediaInfoList) {
            mediaListMap[info.mediaId] = info
        }
    }

    override fun onlyOneMediaBySongInfo(songInfo: SongInfo) {
        val songInfos = ArrayList<SongInfo>()
        songInfos.add(songInfo)
        updateMediaListBySongInfo(songInfos)
    }

    override fun updateMediaListBySongInfo(songInfos: List<SongInfo>) {
        songListMap.clear()
        mMediaMetadataCompatMap.clear()
        songList.clear()
        songList.addAll(songInfos)
        val mediaInfos = mutableListOf<BaseMediaInfo>()
        for (songInfo in songList) {
            val mediaInfo = BaseMediaInfo()
            mediaInfo.mediaId = songInfo.songId
            mediaInfo.mediaTitle = songInfo.songName
            mediaInfo.mediaCover = songInfo.songCover
            mediaInfo.mediaUrl = songInfo.songUrl
            mediaInfo.duration = songInfo.duration
            mediaInfos.add(mediaInfo)
            songListMap[songInfo.songId] = songInfo
        }
        mMediaMetadataCompatMap = toMediaMetadata(songList)
        updateMediaList(mediaInfos)
    }

    override fun addMediaInfo(mediaInfo: BaseMediaInfo?) {
        if (mediaInfo == null) {
            return
        }
        if (!mediaList.contains(mediaInfo)) {
            mediaList.add(mediaInfo)
        }
        mediaListMap[mediaInfo.mediaId] = mediaInfo
    }

    override fun addMediaBySongInfo(info: SongInfo) {
        if (!hasMediaInfo(info.songId)) {
            songList.add(info)
            songListMap[info.songId] = info
            val metadataCompat = toMediaMetadata(info)
            mMediaMetadataCompatMap[info.songId] = metadataCompat

            val mediaInfo = BaseMediaInfo()
            mediaInfo.mediaId = info.songId
            mediaInfo.mediaTitle = info.songName
            mediaInfo.mediaCover = info.songCover
            mediaInfo.mediaUrl = info.songUrl
            mediaInfo.duration = info.duration

            addMediaInfo(mediaInfo)
        }
    }

    override fun deleteMediaById(songId: String) {
        if (hasMediaInfo(songId)) {
            val info = getSongInfo(songId)
            val mediaInfo = getMediaInfo(songId)
            mediaList.remove(mediaInfo)
            mediaListMap.remove(songId)
            songListMap.remove(songId)
            mMediaMetadataCompatMap.remove(songId)
            songList.remove(info)
        }
    }

    override fun hasMediaInfo(songId: String): Boolean {
        return mediaListMap.containsKey(songId)
    }

    override fun getMediaInfo(songId: String): BaseMediaInfo? {
        if (songId.isEmpty()) {
            return null
        }
        return mediaListMap.getOrElse(songId, { null })
    }

    override fun getMediaInfo(index: Int): BaseMediaInfo? {
        return mediaList.elementAtOrNull(index)
    }

    override fun getSongInfo(songId: String): SongInfo? {
        if (TextUtils.isEmpty(songId)) {
            return null
        }
        return songListMap.getOrElse(songId, { null })
    }

    override fun getSongInfo(index: Int): SongInfo? {
        return songList.elementAtOrNull(index)
    }

    override fun getIndexByMediaId(songId: String): Int {
        val info = getMediaInfo(songId)
        return if (info != null) mediaList.indexOf(info) else -1
    }

    override fun getMediaMetadataCompatById(songId: String?): MediaMetadataCompat? {
        if (songId.isNullOrEmpty()) {
            return null
        }
        return mMediaMetadataCompatMap.getOrElse(songId, { null })
    }

    override fun updateMusicArt(
        songId: String, changeData: MediaMetadataCompat, albumArt: Bitmap, icon: Bitmap
    ) {
        val metadata = MediaMetadataCompat.Builder(changeData)
            .putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, albumArt)
            .putBitmap(MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON, icon)
            .build()
        mMediaMetadataCompatMap[songId] = metadata
    }

    /**
     * List<SongInfo> 转 ConcurrentMap<String></String>, MediaMetadataCompat>
    </SongInfo> */
    @Synchronized
    private fun toMediaMetadata(
        songInfos: List<SongInfo>
    ): LinkedHashMap<String, MediaMetadataCompat> {
        val map = linkedMapOf<String, MediaMetadataCompat>()
        for (info in songInfos) {
            val metadataCompat = toMediaMetadata(info)
            map[info.songId] = metadataCompat
        }
        return map
    }

    /**
     * SongInfo 转 MediaMetadataCompat
     */
    @Synchronized
    private fun toMediaMetadata(info: SongInfo): MediaMetadataCompat {
        var albumTitle = ""
        if (!TextUtils.isEmpty(info.albumName)) {
            albumTitle = info.albumName
        } else if (!TextUtils.isEmpty(info.songName)) {
            albumTitle = info.songName
        }
        var songCover = ""
        if (!TextUtils.isEmpty(info.songCover)) {
            songCover = info.songCover
        } else if (!TextUtils.isEmpty(info.albumCover)) {
            songCover = info.albumCover
        }
        val builder = MediaMetadataCompat.Builder()
        builder.id = info.songId
        builder.mediaUri = info.songUrl
        if (albumTitle.isNotEmpty()) {
            builder.album = albumTitle
        }
        if (info.artist.isNotEmpty()) {
            builder.artist = info.artist
        }
        if (info.duration != -1L) {
            builder.duration = info.duration
        }
        if (info.genre.isNotEmpty()) {
            builder.genre = info.genre
        }
        if (songCover.isNotEmpty()) {
            builder.albumArtUri = songCover
        }
        if (info.songName.isNotEmpty()) {
            builder.title = info.songName
        }
        if (info.trackNumber != -1) {
            builder.trackNumber = info.trackNumber.toLong()
        }
        builder.trackCount = info.albumSongCount.toLong()
        return builder.build()
    }
}